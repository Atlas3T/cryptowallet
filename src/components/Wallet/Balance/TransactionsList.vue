<template>
  <section
    class="scroll-area"
    @touchmove="prevent"
  >
    <q-scroll-area
      ref="scrollArea"
      class="scroll-area"
    >
      <q-infinite-scroll
        :handler="loadMore"
        :offset="50"
      >
        <q-timeline
          responsive
          color="secondary"
        >
          <SingleTransaction
            v-for="transaction in filteredPaginated"
            :key="transaction.ts"
            :data="transaction"
          />
        </q-timeline>
        <q-spinner-dots
          slot="message"
          :size="40"
        />
      </q-infinite-scroll>
    </q-scroll-area>
  </section>
</template>

<script>
import SingleTransaction from './SingleTransaction.vue';

export default {
  name: 'TransactionsList',
  components: {
    SingleTransaction,
  },
  props: {
    filter: {
      type: String,
      default: '',
    },
  },
  data() {
    return {
      page: 1,
      perPage: 10,
    };
  },
  computed: {
    transactions() {
      return this.$store.state.payments.transactions;
    },
    /**
     * Filters transactions array from all/sent/received
     * @return {Array}
     */
    filtered() {
      return this.transactions.filter((transaction) => {
        if (this.filter === 'sent') return transaction.to;
        if (this.filter === 'received') return transaction.from;
        return true;
      });
    },
    /**
     * Will slice the filtered transactions array into paginated size
     * @return {Array}
     */
    filteredPaginated() {
      return this.filtered.slice(0, this.page * this.perPage);
    },
  },
  methods: {
    /**
     * Called by q-infinite-scroll component, will attempt to fetch more data
     * The second argument is a function, if said function is called with true
     * it will stop q-infinite-scroll from calling loadMore()
     * @param  {Number} index Current page, begins with 1
     * @param  {Function} done
     * @return {Boolean} true
     */
    async loadMore(index, done) {
      // if you have less data then you could fit on the page
      // it means there is no more data to be loaded
      // call done(true) to prevent subsequent attempts
      if (this.filteredPaginated.length < this.page * this.perPage) {
        done(true);
        return false;
      }

      /**
       * Included to simulate network latency
       * Without it user won't see the loading indicator
       * @param  {Number} ms Milliseconds to wait
       * @return {Function}
       */
      function sleeper(ms) {
        return function (x) {
          return new Promise(resolve => setTimeout(() => resolve(x), ms));
        };
      }

      /**
       * Wait 1500ms and the load the next page
       * await is needed for unit tests to wait for asynchronous code
       */
      await sleeper(500)().then(() => {
        this.page = this.page + 1;
        done(false);
      });

      return true;
    },
    /**
     * Stops q-pull-to-refresh from firing until TransactionsList reaches
     * the top scroll position
     * @param  {Object} event
     */
    prevent(event) {
      if (this.$refs.scrollArea.$el.childNodes[0].scrollTop > 0) {
        event.stopPropagation();
      }
    },
  },
};
</script>

<style>
.scroll-area {
  height: 100%
}
</style>
