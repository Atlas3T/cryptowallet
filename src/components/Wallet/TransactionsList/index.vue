<template>
  <section
    class="scroll-area static"
    @touchstart="touchStart"
    @touchmove="touchMove"
  >
    <q-scroll-area
      ref="scrollArea"
      class="scroll-area static"
    >
      <q-infinite-scroll
        :handler="loadMore"
        :offset="50"
      >
        <q-timeline
          responsive
          color="blueish"
        >
          <SingleTransaction
            v-for="transaction in filteredPaginated"
            :key="transaction.ts"
            :data="transaction"
          />

          <div
            v-if="filteredPaginated.length === 0"
            class="no-tx-alert"
          >
            <q-banner class="info bg-primary">
              {{ $t('emptyTransactionHistory') }}
            </q-banner>
          </div>
        </q-timeline>
        <q-spinner-dots
          slot="message"
          :size="40"
        />
      </q-infinite-scroll>
    </q-scroll-area>
  </section>
</template>

<script>
import SingleTransaction from '@/components/Wallet/SingleTransaction';
import Tx from '@/store/wallet/entities/tx';

export default {
  name: 'TransactionsList',

  components: {
    SingleTransaction,
  },

  props: {
    filter: {
      type: String,
      default: '',
    },
    wallet: {
      type: Object,
      required: true,
    },
  },

  data() {
    return {
      page: 1,
      perPage: 10,
    };
  },
  computed: {
    transactions() {
      const msTos = 1000;
      const txs = Tx.query()
        .where('wallet_id', this.wallet.id)
        .where('isChange', false)
        .get();

      txs.sort((a, b) => {
        if (this.wallet.sdk === 'Ethereum' || this.wallet.sdk === 'ERC20') {
          let b1 = new Date(b.confirmedTime * msTos);
          if (!b.confirmedTime) { b1 = new Date(b.receivedTime.toString() * msTos); }

          let a1 = new Date(a.confirmedTime * msTos);
          if (!a.confirmedTime) { a1 = new Date(a.receivedTime.toString() * msTos); }

          return b1 - a1;
        }

        if (this.wallet.sdk === 'Bitcoin') {
          return new Date(b.receivedTime * msTos) - new Date(a.receivedTime * msTos);
        }

        return 0;
      });

      return txs;
    },

    /**
     * Filters transactions array from all / sent / received
     * @return {Array}
     */
    filtered() {
      return this.transactions.filter((transaction) => {
        if (this.filter === 'sent') { return transaction.sent === true; }
        if (this.filter === 'received') { return transaction.sent === false; }
        return true;
      });
    },

    /**
     * Will slice the filtered transactions array into paginated size
     * @return {Array}
     */

    filteredPaginated() {
      return this.filtered.slice(0, this.page * this.perPage);
    },
  },

  methods: {

    /**
     * add wait to loading tx's
     * @param  {Number} ms Milliseconds to wait
     * @return {Function}
     */
    sleeper(ms) {
      return () => {
        return new Promise((resolve) => {
          return setTimeout(() => {
            return resolve();
          }, ms);
        });
      };
    },

    /**
     * Called by q-infinite-scroll component, will attempt to fetch more data
     * The second argument is a function, if said function is called with true
     * it will stop q-infinite-scroll from calling loadMore()
     * @param  {Number} index Current page, begins with 1
     * @param  {Function} done
     * @return {Boolean} true
     */
    loadMore(index, done) {
      if (this.filteredPaginated.length < this.page * this.perPage) {
        done(true);
        return false;
      }

      return this.sleeper(500)().then(() => {
        this.page = this.page + 1;
        done(false);
      });
    },

    touchStart(event) {
      this.touchStartY = event.touches[0].clientY;
    },

    touchMove(event) {
      if (event.touches[0].clientY <= this.touchStartY) {
        event.stopPropagation();
      }

      if (this.$refs.scrollArea.$el.childNodes[0].scrollTop !== 0) {
        event.stopPropagation();
      }
    },
  },
};
</script>

<style>
.scroll-area {
  height: 100%
}

.scroll-area.static {
  position: static!important;
}

.no-tx-alert {
  padding: 0.5rem;
}

.no-tx-alert .q-alert-content {
  font-size: 0.8rem;
}

.q-tab-panel {
  padding: 0;
}
</style>
