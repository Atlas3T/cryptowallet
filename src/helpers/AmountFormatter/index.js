/**
 * @TODO James Konrad
 * I've mocked some exchange rates below for the amount() function to use
 * Would make sense to put them somewhere in the Vuex store
 */

import numeral from 'numeral';

const exchangeRates = {
  BTC: [
    {
      currency: 'USD',
      rate: 6431.14,
    },
    {
      currency: 'GBP',
      rate: 4977.06,
    },
  ],
  ETH: [
    {
      currency: 'USD',
      rate: 201.76,
    },
    {
      currency: 'GBP',
      rate: 156.14,
    },
  ],
  ATL: [
    {
      currency: 'USD',
      rate: 22222,
    },
    {
      currency: 'GBP',
      rate: 9999,
    },
  ],
};

export default class AmountFormatter {
  constructor(options) {
    this.amount = options.amount;
    this.coin = options.coin;
    this.currency = options.currency;
    this.format = options.format;
    this.withCurrencySymbol = options.withCurrencySymbol;
    this.prependPlusOrMinus = options.prependPlusOrMinus;
  }

  /**
   * Formats a Number and returns a String
   * @return {String}
   */
  getFormatted() {
    let { amount } = this;
    if (this.coin && this.currency) {
      amount = this.coinToCurrency(this.amount, this.coin, this.currency);
    }

    let formatted = `${numeral(Math.abs(amount)).format(this.format)}`;
    if (this.currency) formatted = `${this.currency.symbol}${formatted}`;

    if (this.prependPlusOrMinus) {
      if (amount < 0) return `- ${formatted}`;
      if (amount > 0) return `+ ${formatted}`;
    }

    return formatted;
  }

  /**
   * Converts a coin into a currency
   * @return {Number}
   */
  coinToCurrency() {
    const { rate } = exchangeRates[this.coin].find(item => item.currency === this.currency.code);
    if (rate) return this.amount * rate;
    return this.amount;
  }
}
