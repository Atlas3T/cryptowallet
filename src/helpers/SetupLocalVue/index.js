import Vuex from 'vuex';
import { createLocalVue } from '@vue/test-utils';
import VueRouter from 'vue-router';
import VueI18n from 'vue-i18n';
import messages from '@/i18n/';
import axios from 'axios';
import { sync } from 'vuex-router-sync';
import errorHandlerPlugin from '@/boot/ErrorHandler';
import toasterPlugin from '@/boot/Toaster';
import permissionsPlugin from '@/boot/Permissions';
import appInvitationPlugin from '@/boot/AppInvitation';
import smsPlugin from '@/boot/Sms';
import emailPlugin from '@/boot/Email';
import contactsImportPlugin from '@/boot/ContactsImport';
import * as All from 'quasar'

const { Quasar, date } = All;

const components = Object.keys(All).reduce((object, key) => {
  const val = All[key]
  if (val && val.component && val.component.name != null) {
    object[key] = val
  }
  return object
}, {})


const localVue = createLocalVue();

localVue.use(Quasar, { components });
localVue.use(VueI18n);
localVue.use(Vuex);
// localVue.use(VueRouter);
localVue.prototype.$axios = axios;
errorHandlerPlugin({ Vue: localVue });
toasterPlugin({ Vue: localVue });
permissionsPlugin({ Vue: localVue });
appInvitationPlugin({ Vue: localVue });
smsPlugin({ Vue: localVue });
emailPlugin({ Vue: localVue });
contactsImportPlugin({ Vue: localVue });

/**
 * creates a new router and syncs it with the store using vuex-router-sync
 * @param store
 * @returns router
 */
const createRouter = (store = null) => {
  const router = new VueRouter({
    routes: [
      {
        path: '/contact/:id',
      },
    ],
  });

  if (store) { sync(store, router); }
  return router;
};

const i18n = new VueI18n({ messages, locale: 'en-gb' });

export {
  localVue,
  createRouter,
  i18n,
};
