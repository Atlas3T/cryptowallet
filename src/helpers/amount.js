/**
 * @TODO James Konrad
 * mocked data below
 */

import numeral from 'numeral';

const currencySymbols = {
  USD: '$',
  GPB: 'Â£',
};

const exchangeRates = {
  BTC: [
    {
      currency: 'USD',
      rate: 6431.14,
    },
    {
      currency: 'GPB',
      rate: 4977.06,
    },
  ],
  ETH: [
    {
      currency: 'USD',
      rate: 201.76,
    },
    {
      currency: 'GPB',
      rate: 156.14,
    },
  ],
  ATL: [
    {
      currency: 'USD',
      rate: 22222,
    },
    {
      currency: 'GPB',
      rate: 9999,
    },
  ],
};

export default class Amount {
  constructor(options) {
    this.amount = options.amount;
    this.coin = options.coin;
    this.currency = options.currency;
    this.format = options.format;
    this.withCurrencySymbol = options.withCurrencySymbol;
    this.prependPlusOrMinus = options.prependPlusOrMinus;
  }

  /**
   * Formats a Number and returns a String
   * @return {String}
   */
  get formatted() {
    let { amount } = this;
    if (this.coin && this.currency) {
      amount = this.coinToCurrency(this.amount, this.coin, this.currency);
    }

    let formatted = `${numeral(Math.abs(amount)).format(this.format)}`;
    if (this.currency) formatted = `${currencySymbols[this.currency]}${formatted}`;

    if (this.prependPlusOrMinus) {
      if (amount < 0) return `&#45; ${formatted}`;
      if (amount > 0) return `&#43; ${formatted}`;
    }

    return formatted;
  }

  /**
   * Converts a coin into a currency
   * @return {Number}
   */
  coinToCurrency() {
    const { rate } = exchangeRates[this.coin].find(item => item.currency === this.currency);
    return this.amount * rate;
  }
}
