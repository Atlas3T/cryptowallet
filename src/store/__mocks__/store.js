import Vue from 'vue';
import Vuex from 'vuex';
import VuexORM from '@vuex-orm/core';
import VuexORMLoki from '@atlascity/vuex-orm-plugin-lokijs';

import wallet from '@/store/wallet';
import contacts from '@/store/contacts/__mocks__/contacts.js';
import search from '@/store/search/__mocks__/search.js';
import settings from '@/store/settings/__mocks__/settings.js';
import setup from '@/store/setup/__mocks__/setup.js';
import qrcode from '@/store/qrcode/__mocks__/qrcode.js';
import modals from '@/store/modals/__mocks__/modals.js';

// entities
import Account from '@/store/wallet/entities/account';
import Address from '@/store/wallet/entities/address';
import Tx from '@/store/wallet/entities/tx';
import Utxo from '@/store/wallet/entities/utxo';
import Wallet from '@/store/wallet/entities/wallet';
import Coin from '@/store/wallet/entities/coin';
import KeyPair from '@/store/wallet/entities/keyPair';
import Prices from '@/store/prices';
import LatestPrice from '@/store/latestPrice';

Vue.use(Vuex);

// Setup ORM database.
const database = new VuexORM.Database();

database.register(Account, {});
database.register(Address, {});
database.register(Tx, {});
database.register(Utxo, {});
database.register(Wallet, {});
database.register(Coin, {});
database.register(KeyPair, {});
database.register(Prices, {});
database.register(LatestPrice, {});

Vue.prototype.encryptedModels = [
  Account,
  Wallet,
];

const options = {
  env: 'browser',
};

VuexORM.use(VuexORMLoki, { database, options });

export const actions = {};
export const getters = {};
export const mutations = {};
export const state = {};

export function createMocks(custom = {
  getters: {}, mutations: {}, actions: {}, state: {},
}) {
  /**
   * mock store modules.
   */
  const modules = {
    wallet,
    search,
    settings,
    contacts,
    setup,
    qrcode,
    modals,
  };

  let mockGettersCombined = {};
  let mockMutationsCombined = {};
  let mockActionsCombined = {};
  let mockStateCombined = {};

  Object.entries(modules).forEach(([name, module]) => {
    mockGettersCombined = Object.assign({}, mockGettersCombined, module.getters);
    mockMutationsCombined = Object.assign({}, mockMutationsCombined, module.mutations);
    mockActionsCombined = Object.assign({}, mockActionsCombined, module.actions);
    mockStateCombined = Object.assign({}, mockStateCombined, module.state);

    if (custom.state && custom.state[name]) {
      module.state = Object.assign(module.state, custom.state[name]);
    }

    modules[name] = {
      namespaced: true,
      getters: Object.assign({}, getters, module.getters),
      mutations: Object.assign({}, mutations, module.mutations),
      actions: Object.assign({}, actions, module.actions),
      state: Object.assign({}, state, module.state),
    };
  });

  mockGettersCombined = Object.assign({}, mockGettersCombined, custom.getters);
  mockMutationsCombined = Object.assign({}, mockMutationsCombined, custom.mutations);
  mockActionsCombined = Object.assign({}, mockActionsCombined, custom.actions);
  mockStateCombined = Object.assign({}, mockStateCombined, custom.state);
  return {
    getters: mockGettersCombined,
    mutations: mockMutationsCombined,
    actions: mockActionsCombined,
    state: mockStateCombined,
    store: new Vuex.Store({
      getters: mockGettersCombined,
      mutations: mockMutationsCombined,
      actions: mockActionsCombined,
      state: mockStateCombined,
      modules,
      plugins: [VuexORM.install(database)],
    }),
  };
}

export const { store } = createMocks();
