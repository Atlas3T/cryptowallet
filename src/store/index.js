/*eslint-disable*/

import Vue from 'vue';
import Vuex from 'vuex';
import VuexPersistence from 'vuex-persist';
import VuexORM from '@vuex-orm/core';
import VuexORMLoki from 'vuex-orm-lokijs';

// import entities.
import Account from './wallet/entities/account';
import Address from './wallet/entities/address';
import Tx from './wallet/entities/tx';
import Utxo from './wallet/entities/utxo';
import Wallet from './wallet/entities/wallet';
import Coin from './wallet/entities/coin';
import KeyPair from './wallet/entities/keyPair';
import Prices from './prices';
import LatestPrice from './latestPrice';

// import modules.
import payments from './payments';
import search from './search';
import settings from './settings';
import setup from './setup';
import contacts from './contacts';

Vue.use(Vuex);

// Setup ORM database.
const database = new VuexORM.Database();

database.register(Account, {});
database.register(Address, {});
database.register(Tx, {});
database.register(Utxo, {});
database.register(Wallet, {});
database.register(Coin, {});
database.register(KeyPair, {});
database.register(Prices, {});
database.register(LatestPrice, {});

Vue.prototype.encryptedModels = [
  Account,
];

window.Tx = Tx;
window.Utxo = Utxo;
window.Address = Address;
window.Wallet = Wallet;
window.LatestPrice = LatestPrice;
window.Prices = Prices;
window.Account = Account;
window.Coin = Coin;

// Setup persistant storage.
const vuexLocal = new VuexPersistence({
  supportCircular: true,
  storage: window.localStorage,
  modules: [settings],
});


const options = {
  env: 'browser',
};

function hydrationCompletedCallback() {
  setTimeout(() => {
    store.dispatch('settings/setLoading', false);
  }, 1000);
}

VuexORM.use(VuexORMLoki, { database, options, hydrationCompletedCallback });


/**
 * Create CryptoWallet Vuex store obj.
 * @type {Store<any>}
 */
const store = new Vuex.Store({
  modules: {
    payments,
    search,
    settings,
    setup,
    contacts,
  },
  plugins: [VuexORM.install(database)],
});

Vue.prototype.activeWallets = {};

// @todo remove line below (used for development purposes only)
window.store = store;

export default store;
