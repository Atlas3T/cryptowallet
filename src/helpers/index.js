import AmountFormatter from './AmountFormatter';
import dateTranslater from './DateTranslater';
import Contacts from './Contacts';

export {
  AmountFormatter,
  dateTranslater,
  Contacts,
};
