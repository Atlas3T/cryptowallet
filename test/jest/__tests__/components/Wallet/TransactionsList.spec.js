/* eslint-disable no-magic-numbers */
import { shallowMount } from '@vue/test-utils';
import TransactionsList from '@/components/Wallet/TransactionsList';
import { localVue, i18n, createRouter } from '@/helpers/SetupLocalVue';
import { createMocks as createStoreMocks } from '@/store/__mocks__/store.js';
import {
  mockBitcoinTxs, mockBitcoinTxsSorted, mockEthereumTxs, mockEthereumTxsSorted,
} from '@/store/wallet/__mocks__/tx.js';
import Tx from '@/store/wallet/entities/tx';

describe('TransactionsList.vue', () => {
  let storeMocks;
  let wrapper;
  // let store;
  let router;

  const propsData = {
    filter: 'all',
    wallet: {
      id: 4,
      account_id: 1,
      displayName: 'Bitcoin',
      sdk: 'Bitcoin',
    },
  };

  function wrapperInit(options, empty) {
    if (!empty) {
      Tx.insert({ data: mockBitcoinTxs.data });
    }
    return shallowMount(TransactionsList, options);
  }

  function storeInit(empty = false) {
    storeMocks = createStoreMocks();
    router = createRouter(storeMocks.store);
    router.push({ path: '/' });
    wrapper = wrapperInit({
      i18n, router, localVue, store: storeMocks.store, propsData,
    }, empty);
    // store = wrapper.vm.$store;
  }

  beforeEach(() => { return storeInit(); });

  it('renders and matches snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders a section with .scroll-area class', () => {
    expect(wrapper.contains('section.scroll-area')).toBe(true);
  });

  it('computes Bitcoin transactions correctly', () => {
    expect(wrapper.vm.transactions).toEqual(mockBitcoinTxsSorted.data);
  });

  it('computes Ethereum transactions correctly', () => {
    Tx.insert({ data: mockEthereumTxs.data });
    wrapper.setProps({
      filter: 'all',
      wallet: {
        id: 5,
        account_id: 1,
        displayName: 'Ethereum',
        sdk: 'Ethereum',
      },
    });
    expect(wrapper.vm.transactions).toEqual(mockEthereumTxsSorted.data);
  });

  it('displays a q-banner if there are no transactions', () => {
    storeInit(true);
    expect(wrapper.contains('q-banner-stub')).toBe(true);
    expect(wrapper.text()).toBe('Transaction history is empty.');
  });

  it('computes filtered correctly', () => {
    expect(wrapper.vm.filtered.length).toBe(22);

    wrapper.setProps({ filter: 'sent' });
    expect(wrapper.vm.filtered.length).toBe(12);
    wrapper.vm.filtered.forEach((tx) => {
      expect(tx.sent).toBeTruthy();
    });

    wrapper.setProps({ filter: 'received' });
    expect(wrapper.vm.filtered.length).toBe(10);
    wrapper.vm.filtered.forEach((tx) => {
      expect(tx.sent).toBeFalsy();
    });
  });

  it('computes filteredPaginated correctly', () => {
    expect(wrapper.vm.filteredPaginated.length).toBe(10);

    wrapper.setData({
      page: 2,
      perPage: 10,
    });
    expect(wrapper.vm.filteredPaginated.length).toBe(20);

    wrapper.setData({
      page: 2,
      perPage: 6,
    });
    expect(wrapper.vm.filteredPaginated.length).toBe(12);
  });

  it('calls done() correctly within loadMore() method', async () => {
    wrapper.setData({
      page: 1,
      perPage: 9999,
    });
    const doneMock = jest.fn();
    wrapper.vm.loadMore(null, doneMock);
    expect(doneMock).toHaveBeenLastCalledWith(true);

    wrapper.setData({
      page: 1,
      perPage: 1,
    });
    await wrapper.vm.loadMore(null, doneMock);
    expect(doneMock).toHaveBeenLastCalledWith(false);
  });

  it('calls the touchStart() method on touchstart event', () => {
    wrapper.find('section.scroll-area').trigger('touchstart', { touches: [{ clientY: 399 }] });
    expect(wrapper.vm.touchStartY).toBe(399);
  });

  it('calls the touchMove() method on touchMove event', () => {
    const mockStopPropagation = jest.fn();
    wrapper.vm.$refs.scrollArea.$el.childNodes[0].scrollTop = 0;
    wrapper.find('section.scroll-area').trigger('touchstart', { touches: [{ clientY: 399 }] });
    wrapper.find('section.scroll-area').trigger('touchmove', { touches: [{ clientY: 400 }] });
    expect(wrapper.vm.touchStartY).toBe(399);
    wrapper.vm.$refs.scrollArea.$el.childNodes[0].scrollTop = 10;
    wrapper.find('section.scroll-area').trigger('touchstart', { touches: [{ clientY: 399 }] });
    wrapper.find('section.scroll-area').trigger('touchmove', { stopPropagation: mockStopPropagation, touches: [{ clientY: 399 }] });
    expect(mockStopPropagation).toHaveBeenCalledTimes(2);
  });

  //   it('calls event.stopPropagation() correctly within prevent() method', () => {
  //     const event = { stopPropagation: jest.fn() };
  //     wrapper.vm.$refs.scrollArea.$el.childNodes[0].scrollTop = 0;
  //     wrapper.vm.prevent(event);
  //     expect(event.stopPropagation).toHaveBeenCalledTimes(0);

//     wrapper.vm.$refs.scrollArea.$el.childNodes[0].scrollTop = 10;
//     wrapper.vm.prevent(event);
//     expect(event.stopPropagation).toHaveBeenCalledTimes(1);
//   });
});
