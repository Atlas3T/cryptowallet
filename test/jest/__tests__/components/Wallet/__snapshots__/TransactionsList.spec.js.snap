// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionsList.vue renders and matches snapshot 1`] = `
<section
  class="scroll-area static"
>
  <q-scroll-area-stub
    class="scroll-area static"
    contentactivestyle="[object Object]"
    contentstyle="[object Object]"
    delay="1000"
    thumbstyle="[object Object]"
  >
    <q-infinite-scroll-stub
      handler="function () { [native code] }"
      offset="50"
    >
      <q-timeline-stub
        color="blueish"
        layout="dense"
        responsive=""
        side="right"
      >
        <singletransaction-stub
          data="[object Object]"
        />
        <singletransaction-stub
          data="[object Object]"
        />
        <singletransaction-stub
          data="[object Object]"
        />
        <singletransaction-stub
          data="[object Object]"
        />
        <singletransaction-stub
          data="[object Object]"
        />
        <singletransaction-stub
          data="[object Object]"
        />
        <singletransaction-stub
          data="[object Object]"
        />
        <singletransaction-stub
          data="[object Object]"
        />
        <singletransaction-stub
          data="[object Object]"
        />
        <singletransaction-stub
          data="[object Object]"
        />
         
        <!---->
      </q-timeline-stub>
       
      <q-spinner-dots-stub
        size="40"
      />
    </q-infinite-scroll-stub>
  </q-scroll-area-stub>
</section>
`;
